worker/src/main/scala/com/lucidchart/piezo/ConnectionProvider.scala

```diff
@@ -1,25 +1,162 @@
+   * https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/blue-green-deployments-switching.html#blue-green-deployments-switching-actions
+   * When that happens, the hikari pool
+   *   - Automatically closes the connections when they were closed remotely by AWS
+   *   - Automatically attempts to open up new connections to maintain the minimum # of idle connections. There is a
+   *     timeout for how long it waits for a connection to be established. If those connections timeout/get rejected by
+   *     AWS because the switchover isn't complete, the pool keeps attempting to open up new connections until the
+   *     minimum # is reached
+   *
+   * This function returns True, iff
+   *   - IP addresses have changed for the CNAME record used for DNS lookup
+   *
+   * @param pool
+   *   the connection pool currently being used
+   * @param dnsIP
+   *   the IP returned when performing a DNS lookup
+   * @return
+   */
+  private def mustKillConnectionPool(pool: Pool, dnsIP: String): Boolean = {
+    pool.ip != dnsip
+  }
+
+  /**
+   * Resolve the hostname to an ip address. Uses the built-in JVM cache but if that throws an exception, uses our cache
+   * with the last JVM value
+   *
+   * @param hostname - Comes from the config (which is set by lucidbag), which is a CNAME to a writer endpoint. Failovers for writer nodes
+   *                 are caused by a change in IP addresses, so a DNS lookup of the IP address will help us identify when a failover occurred
+   */
+  def getIP: String = {
+    Try {
+      // Get the ip address of the hostname.  The result is cached in the JVM
+      java.net.InetAddress.getByName(dataSourceHostname).getHostAddress
+    } match {
+      case Failure(e) => {
+        // Get the last cached version
+        Option(dnsCache.get(dataSourceHostname)).asInstanceOf[String] match {
+          case Some(ip) => {
+            ip
+          }
+          case None => {
+            // We failed to resolve it from JVM and did not have a cached version,
+            // so throw the exception
+            throw e
+          }
+        }
+      }
+      case Success(ip) => {
+        // Update the DnsCache, so if DNS fails later we have a backup.
+        dnsCache.put(dataSourceHostname, ip)
+        ip
+      }
+    }
   }
 
-  def getConnection = provider.get.getConnection
+  def getConnection = {
+    if(dataSource != null) {
+      // If the IP has changed, then we know a failover has occurred, and we need to create a new hikari config
+      val newIP: String = getIP
+      if (mustKillConnectionPool(pool, newIP)) {
+        // A failover has occurred, so we lock the pool and swap it out with a new hikari config
+        val (poolToUse, optionalOldPool) = poolLock synchronized {
+          val oldPool = pool
+          // If the IP has changed, then we know a failover has occurred
+          val newIP: String = getIP
+          if (mustKillConnectionPool(oldPool, newIP)) {
```

Comment:
If `getIp` doesn't always return the same ip, as would be the case if you have multiple nodes and are load balancing with DNS (more likely for read connections, but could also be the case for master-master replication scenarios), then this will constantly churn between pools for the different ip addresses.

Reply:
That is true. The intention is to support systems in which an IP change would necessitate closing/reopening connections. I think an additional configuration option for using IP-based failover probably makes sense here